File: lulu/binary_tree_paths.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    # @param {TreeNode} root the root of the binary tree
    # @return {List[str]} all root-to-leaf paths
    def binaryTreePaths(self, root):
        # Write your code here
        if root is None:
            return []
        paths = []
        path = ''
        self.find_path(root, paths, path)
        return paths


    def find_path(self, root, paths, path):
        if root.left is None and root.right is None:
            path += str(root.val)
            paths.append(path)
        else:
            path += str(root.val) + '->'
            if root.left is not None:
                self.find_path(root.left, paths, path)
            if root.right is not None:
                self.find_path(root.right, paths, path)

File: bsdelf/binary-tree-paths.cc
/**
 * @author      Yanhui Shen
 * @version     1.0
 * @since       2016-02-11 
 */

/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param root the root of the binary tree
     * @return all root-to-leaf paths
     */
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> paths;
        const std::function<void (TreeNode*, string)> walk = 
            [&walk, &paths](TreeNode* node, string path) {
                if (not node) return;
                path += std::to_string(node->val);
                if (node->left || node->right) {
                    path += "->";
                } else {
                    paths.emplace_back(std::move(path));
                }
                walk(node->left, path);
                walk(node->right, path);
            };
        walk(root, "");
        return paths;
    }
};
